File: src/comp/util/bits.rs
import std._uint;
import std._int;
import std._vec;

// FIXME: With recursive object types, we could implement binary methods like
//        union, intersection, and difference. At that point, we could write
//        an optimizing version of this module that produces a different obj
//        for the case where nbits < 32.

state type t = rec(vec[mutable uint] storage, uint nbits);

// FIXME: this should be a constant once they work
fn uint_bits() -> uint {
    ret 32u + ((1u << 32u) >> 27u) - 1u;
}

// FIXME: this should be state
fn create(uint nbits, bool init) -> t {
    auto elt;
    if (init) {
        elt = 1u;
    } else {
        elt = 0u;
    }

    ret rec(storage = _vec.init_elt[mutable uint](nbits / uint_bits() + 1u, elt),
            nbits = nbits);
}

// FIXME: this should be state
fn process(fn(uint, uint) -> uint op, t v0, t v1) -> bool {
    auto len = _vec.len[mutable uint](v1.storage);

    check (_vec.len[mutable uint](v0.storage) == len);
    check (v0.nbits == v1.nbits);

    auto changed = false;

    for each (uint i in _uint.range(0u, len)) {
        auto w0 = v0.storage.(i);
        auto w1 = v1.storage.(i);

        auto w = op(w0, w1);
        if (w0 != w) {
            changed = true;
            v0.storage.(i) = w;
        }
    }

    ret changed;
}

fn lor(uint w0, uint w1) -> uint {
    ret w0 | w1;
}

// FIXME: this should be state
fn union(t v0, t v1) -> bool {
    auto sub = lor;
    ret process(sub, v0, v1);
}

fn land(uint w0, uint w1) -> uint {
    ret w0 & w1;
}

// FIXME: this should be state
fn intersect(t v0, t v1) -> bool {
    auto sub = land;
    ret process(sub, v0, v1);
}

fn right(uint w0, uint w1) -> uint {
    ret w1;
}

// FIXME: this should be state
fn copy(t v0, t v1) -> bool {
    auto sub = right;
    ret process(sub, v0, v1);
}

// FIXME: this should be state
fn get(t v, uint i) -> bool {
    check (i < v.nbits);

    auto bits = uint_bits();

    auto w = i / bits;
    auto b = i % bits;
    auto x = 1u & (v.storage.(w) >> b);
    ret x == 1u;
}

// FIXME: this should be state
fn equal(t v0, t v1) -> bool {
    // FIXME: when we can break or return from inside an iterator loop,
    //        we can eliminate this painful while-loop
    auto len = _vec.len[mutable uint](v1.storage);
    auto i = 0u;
    while (i < len) {
        if (v0.storage.(i) != v1.storage.(i)) {
            ret false;
        }
        i = i + 1u;
    }
    ret true;
}

// FIXME: this should be state
fn clear(t v) {
    for each (uint i in _uint.range(0u, _vec.len[mutable uint](v.storage))) {
        v.storage.(i) = 0u;
    }
}

// FIXME: this should be state
fn invert(t v) {
    for each (uint i in _uint.range(0u, _vec.len[mutable uint](v.storage))) {
        v.storage.(i) = ~v.storage.(i);
    }
}

// FIXME: this should be state
/* v0 = v0 - v1 */
fn difference(t v0, t v1) -> bool {
    invert(v1);
    auto b = intersect(v0, v1);
    invert(v1);
    ret b;
}

// FIXME: this should be state
fn set(t v, uint i, bool x) {
    check (i < v.nbits);

    auto bits = uint_bits();

    auto w = i / bits;
    auto b = i % bits;
    auto w0 = v.storage.(w);
    auto flag = 1u << b;
    if (x) {
        v.storage.(w) = v.storage.(w) | flag;
    } else {
        v.storage.(w) = v.storage.(w) & ~flag;
    }
}

// FIXME: this should be state
fn init_to_vec(t v, uint i) -> uint {
    if (get(v, i)) {
        ret 1u;
    } else {
        ret 0u;
    }
}

// FIXME: this should be state
fn to_vec(t v) -> vec[uint] {
    auto sub = bind init_to_vec(v, _);
    ret _vec.init_fn[uint](sub, v.nbits);
}

//
// Local Variables:
// mode: rust
// fill-column: 78;
// indent-tabs-mode: nil
// c-basic-offset: 4
// buffer-file-coding-system: utf-8-unix
// compile-command: "make -k -C ../.. 2>&1 | sed -e 's/\\/x\\//x:\\//g'";
// End:
//

File: src/boot/util/bits.ml
type t = {
  storage: int array;
  nbits: int;
}
;;

let int_bits =
  if max_int = (1 lsl 30) - 1
  then 31
  else 63
;;

let create nbits flag =
  { storage = Array.make (nbits / int_bits + 1) (if flag then lnot 0 else 0);
    nbits = nbits }
;;

(* 
 * mutate v0 in place: v0.(i) <- v0.(i) op v1.(i), returning bool indicating
 * whether any bits in v0 changed in the process. 
 *)
let process (op:int -> int -> int) (v0:t) (v1:t) : bool =
  let changed = ref false in
    assert (v0.nbits = v1.nbits);
    assert ((Array.length v0.storage) = (Array.length v1.storage));
    Array.iteri
      begin
        fun i w1 ->
          let w0 = v0.storage.(i) in
          let w0' = op w0 w1 in
            if not (w0' = w0)
            then changed := true;
            v0.storage.(i) <- w0';
      end
      v1.storage;
    !changed
;;

let union = process (lor) ;;
let intersect = process (land) ;;
let copy = process (fun _ w1 -> w1) ;;

let get (v:t) (i:int) : bool =
  assert (i >= 0);
  assert (i < v.nbits);
  let w = i / int_bits in
  let b = i mod int_bits in
  let x = 1 land (v.storage.(w) lsr b) in
    x = 1
;;

let equal (v1:t) (v0:t) : bool =
  v0 = v1
;;

let clear (v:t) : unit =
  for i = 0 to (Array.length v.storage) - 1
  do
    v.storage.(i) <- 0
  done
;;

let invert (v:t) : unit =
  for i = 0 to (Array.length v.storage) - 1
  do
    v.storage.(i) <- lnot v.storage.(i)
  done
;;

(* dst = dst - src *)
let difference (dst:t) (src:t) : bool =
  invert src;
  let b = intersect dst src in
    invert src;
    b
;;


let set (v:t) (i:int) (x:bool) : unit =
  assert (i >= 0);
  assert (i < v.nbits);
  let w = i / int_bits in
  let b = i mod int_bits in
  let w0 = v.storage.(w) in
  let flag = 1 lsl b in
    v.storage.(w) <-
      if x
      then w0 lor flag
      else w0 land (lnot flag)
;;

let to_list (v:t) : int list =
  if v.nbits = 0
  then []
  else
    let accum = ref [] in
    let word = ref v.storage.(0) in
      for i = 0 to (v.nbits-1) do
        if i mod int_bits = 0
        then word := v.storage.(i / int_bits);
        if (1 land (!word)) = 1
        then accum := i :: (!accum);
        word := (!word) lsr 1;
      done;
      !accum
;;


(*
 * Local Variables:
 * fill-column: 78;
 * indent-tabs-mode: nil
 * buffer-file-coding-system: utf-8-unix
 * compile-command: "make -k -C $RBUILD 2>&1 | sed -e 's/\\/x\\//x:\\//g'";
 * End:
 *)
